### PR TITLE
Refactor RequestStream rate limit strategies

### PR DESCRIPTION
**Pull request by quanticc**
**Description:** Improve add reaction ratelimit strategy and its consequences
**Justification:** RequestStream needed more love from us

Summary of changes:
1. Move GatewayLimiter interface to RateLimiter in common module (because now rest uses it)
2. Extract current header-only rate limit strategy to support the
RateLimiter interface
3. Make Routes.REACTION_CREATE use a separate strategy backed by a
SimpleBucket, allowing undocumented but "expected" behavior
4. Include an example command handler to test this capability (who wants to refactor those emojis, maybe unicode range, or a separate library?)
5. Rework GlobalRateLimiter and the main RequestStream chain, signal
routing works the same, add error handling to quickly detect
potential errors in the implementation
6. Forward malformed client requests back to the user as proper
onError signals (#485) this is just a bonus from refactoring the whole RequestStream chain. Thanks jarza#1289 for letting us know.

### CODE REVIEW

File: core/src/test/java/discord4j/core/ExampleBot.java [hunk 1]
```diff
@@ -186,8 +191,90 @@ public void testCommandBot() {
...
+            return emojiList;
+        }
+    }
+
     public static abstract class EventHandler {
 
         public abstract Mono<Void> onMessageCreate(MessageCreateEvent event);
     }
+
+    private static Predicate<Throwable> isStatusCode(int code) {
```
[the remark this answers is not in the document]
Comment (quanticc) [reply core/src/test/java/discord4j/core/ExampleBot.java]: Not yet, but we should discuss that separately, because there are also plans to provide a cross-cutting error strategy: https://github.com/Discord4J/Discord4J/compare/feature/on-status?expand=1

[the remark this answers is not in the document]
Comment (quanticc) [reply core/src/test/java/discord4j/core/ExampleBot.java]: See above comment

File: core/src/test/java/discord4j/core/ExampleBot.java [hunk 2]
```diff
@@ -186,8 +191,90 @@ public void testCommandBot() {
         }
     }
 
+    public static class Reactor extends EventHandler {
+
+        private final Random random = new Random();
+        // yes this is going into our codebase thanks
+        private final String[] emoji = {"😀", "😬", "😂", "😄", "😅", "😇", "☺️", "😋", "😘", "😚", "😜", "🤑", "😎"
```
[the remark this answers is not in the document]
Comment (quanticc) [reply core/src/test/java/discord4j/core/ExampleBot.java]: Do you mean the maximum of 10? Not really, it's just that the code is too naive and duplicates will happen. I think I could refactor it.

File: common/src/test/java/discord4j/common/SimpleBucketTest.java [hunk 3]
```diff
@@ -39,7 +39,7 @@
     @Ignore
     public void testReactiveBucket() throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (quanticc) [reply common/src/test/java/discord4j/common/SimpleBucketTest.java]: You're right, let me attempt moving to StepVerifier. Need to learn more about it

**Comment by quanticc** [comment]
@GrandPanda We have a test on rest module, reacts with ❤️  https://github.com/Discord4J/Discord4J/blob/v3/rest/src/test/java/discord4j/rest/service/ChannelServiceTest.java#L133
